Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: MultipleVariableDeclarations

child of #14942 

check doc:  https://checkstyle.org/checks/coding/multiplevariabledeclarations.html



---

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="MultipleVariableDeclarationsCheck">
        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
import java.util.LinkedHashMap;
import java.util.List;

record ColoredPoint(String p, String x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
       int x = 0, _ = sideEffect();

    }
    int sideEffect() {
        return 0;
    }
}

PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:8: Each variable declaration must be in its own statement. [MultipleVariableDeclarations]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```

---

The check works as intended.

Yes.

```
PS D:\CS\test> cat src/Test.java                                                
public class Test {
    void test(Object obj) {
        int _ = 0, _ = sideEffect();  // violation
    }
    int sideEffect() { return 0;    }
}
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
            <module name="MultipleVariableDeclarations"/>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:9: Each variable declaration must be in its own statement. [MultipleVariableDeclarations]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MultipleVariableDeclarationsCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/MultipleVariableDeclarationsCheckTest.java
@@ -81,4 +81,19 @@ public class MultipleVariableDeclarationsCheckTest extends AbstractModuleTestSup
                expected);
     }
 
+    @Test
+    public void testUnnamedVariables() throws Exception {
+
+        final String[] expected = {
+            "12:9: " + getCheckMessage(MSG_MULTIPLE_COMMA),
+            "14:9: " + getCheckMessage(MSG_MULTIPLE_COMMA),
+            "16:9: " + getCheckMessage(MSG_MULTIPLE),
+            "18:9: " + getCheckMessage(MSG_MULTIPLE),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputMultipleVariableDeclarationsUnnamedVariables.java"),
+               expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/multiplevariabledeclarations/InputMultipleVariableDeclarationsUnnamedVariables.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/multiplevariabledeclarations/InputMultipleVariableDeclarationsUnnamedVariables.java
@@ -1,0 +1,24 @@
+/*
+MultipleVariableDeclarations
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.multiplevariabledeclarations;
+
+public class InputMultipleVariableDeclarationsUnnamedVariables {
+    void test() {
+        // violation below, 'Each variable declaration must be in its own statement'
+        int _ = sideEffect(), _ = sideEffect();
+        // violation below, 'Each variable declaration must be in its own statement'
+        int _ = sideEffect(), a = sideEffect();
+        // violation below, 'Only one variable definition per line allowed'
+        int _ = sideEffect(); int _ = sideEffect();
+        // violation below, 'Only one variable definition per line allowed'
+        int _ = sideEffect(); int b = sideEffect();
+    }
+
+    int sideEffect() {
+        return 0;
+    }
+}

